### PR TITLE
fix: circular dependencies

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [ ] I branched from develop
- [ ] My pull request is against develop
- [ ] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [ ] I ran `npm run format` and `npm run lint`

### Resolves

Built code not actually working, due to circular dependencies.

### Proposed Changes

Does various things to make the built code actually work.
  * Moves `getById` and `getAll` into common to break some dependency cycles.
  * Comments out imports of certain things being used for side effects to break dependency cycles. Basically I commented them out until it started working.
  * Creates a bunch of "stub" functions/methods that do not have implementations in the file they belong to (so that we can break dependency cycles). The implementations are monkey-patched in in blockly.ts
  * Creates internal versions of a bunch of functions that are added to the `TEST_ONLY` export of their respective modules so that they can be spied/stubbed in tests.
  * Changes `goog.js` to include actual definitions of its functions, instead of pointing to the global `goog` and assuming it exists. The closure compiler couldn't figure out how to order things to get this to work. The definitions in `goog.js` just call the global `goog`.
  * Commented out the `declareModuleId` error we were getting in uncompiled mode (using the debug module loader), so that we can run tests and the playground.

#### Behavior Before Change

Built code did not work.

#### Behavior After Change

Built code works!

### Reason for Changes

Things working is good.

### Documentation

N/A

### Additional Information

This is super messy because I was just trying to get things to work. Happy to clean up later, but wanted to put this up for discussion. Also happy to answer any questions about the weird things I am doing.

### Future work

* Investigate whether those commented-out imports are necessary or not.
* Get rid of any internal functions that were created for test stubbing.
* Figure out a way to run tests without commenting out the `declareModuleId` error. Eg use webpack or ESModule loading or something.

### CODE REVIEW

File: closure/goog/goog.js
Comment (cpcallen) [inline closure/goog/goog.js:134]: Like `goog.require`, ``goog.declareModuleId` is is never needed in compiled mode—the compiler treats the call as declaration/pragma, and emits no code for it.  Unlike `require`, however, we (for now) definitely have calls to `declareModuleId` that actually have to happen in uncompiled mode.  But—contrary to my previous advice—I think it would be preferable to farm them out to the real implementation, if `tsc` and Closure Compiler will permit us:

```suggestion
export const declareModuleId = function(namespace) {
  if (window.goog && window.goog.declareModuleId) {
    window.goog.require.call(this, namespace);
  }
};
```

(Or `…(window.goog, namespace)…` if anything complains about the use of `this`.)

File: closure/goog/goog.js
Comment (cpcallen) [inline closure/goog/goog.js:141]: You know… 🧐

One thing we haven't tried but could try as an alternative to the present PR is just to reintroduce the `goog.module.get` calls that were previously replaced with `import`s.  I'm not sure how happy `tsc` would be, but in principle that _should_ solve all our our circular dependency problems.  🤔🤔

File: core/procedures.ts
Comment (cpcallen) [inline core/procedures.ts:359]: Line length.

File: core/utils/xml.ts
Comment (cpcallen) [inline core/utils/xml.ts:106]: Missing final `\n`.

File: core/workspace_svg.ts
Comment (cpcallen) [inline core/workspace_svg.ts:966]:
```suggestion
  static newTrashcan(workspace: WorkspaceSvg): Trashcan {
    throw new Error('unimplemented');
```

File: closure/goog/goog.js
Comment (BeksOmega) [reply closure/goog/goog.js:64]: I don't think we use it in any of our TS files, but we do use it in the blocks and the generators, which are also compiled using the `goog.js` file.  SO I think we need to keep it? But I'll give removing it a shot as well.

File: closure/goog/goog.js
Comment (BeksOmega) [reply closure/goog/goog.js:141]: Attempted this, and it built and ran, but it did not fix the uncompiled problems (ie mocha tests) so we probably won't dedicate more time to this solution.

File: core/utils/xml.ts
Comment (BeksOmega) [reply core/utils/xml.ts:99]: Breaks a circular dependency between `xml.ts` and `variables.ts`

File: closure/goog/goog.js
Comment (BeksOmega) [reply closure/goog/goog.js:64]: Oh I guess the thing is the blocks use the global `goog`? Not the one imported from the `goog.js` module?

File: core/procedures.ts
Comment (BeksOmega) [reply core/procedures.ts:338]: At some point my IDE was complaining about insufficient type overlap, but it seems that it doesn't care anymore.

File: core/variables.ts
Comment (BeksOmega) [reply core/variables.ts:178]: Yes, so I guess this change wasn't actually necessary to get the build to work. Will see what happens when I revert it.

Comment (BeksOmega) [reply core/variables.ts]: Seems like it was unnecessary! Thanks for the catch :D

File: closure/goog/goog.js
Comment (cpcallen) [reply closure/goog/goog.js]: Correct: blocks and generators have not yet been migrated to ES/TS modules, so do not use `goog.js`, only the versions defined in `base.js`.

File: core/utils/xml.ts
Comment (cpcallen) [reply core/utils/xml.ts:97]: Not your fault, but still missing a newline at the end of the file…

File: core/blockly.ts
Comment (cpcallen) [inline core/blockly.ts:585]: Go home clang-format, you're drunk.  (Line 581: WTF?)

Could the tool be persuaded to be slightly less toolish by manually inserting a newline after the `=` in both declarations?  Or… well, please: just see what you can do to mitigate this fiasco.

**Review by cpcallen — APPROVED** [review]
Approved but with nits.

**Comment by cpcallen** [comment]
I will leave it to your judgement about whether you think this or the `goog.module.get` approach is the more promising one to go with.

File: core/procedures.ts
Comment (cpcallen) [reply core/procedures.ts]: "complaining about insufficient type overlap" smells like a bug (albeit possibly in the editor rather than the code).

**Review by cpcallen — APPROVED** [review]
Test definitely run!  (Not pass… but very much a step in the right direction.)

**Review by maribethb — APPROVED** [review]
Overall thoughts

1) I don't love the patching approach, but it looks like that is what is going to unblock the circular dependency problem and allow us a path forward, so I can get over it.
2) I strongly dislike the TEST_ONLY solution to the mocking problem :( If this is the best path for unblocking us, then I can get over it for now but I really don't want this to stick around. I looked into it and it looks like there are some solutions for mocking modules like proxyquire, quibble, etc that allow us to basically sub in a fake module that we can stub. Alternatively I think if the tests used the compiled code we wouldn't run into this issue? Or some tests could be rearchitected to not have to stub as many things. Or if we make bigger changes to Blockly's architecture to fix some of the dependency cycles then we could also make some changes to allow for easier testing. I think in the near future after this is landed we should devote some time to really dig into current best practices for tests and improve our story on that front.
3) Can you update the PR description to add more information, since in the future this is likely where we'll trace these changes to. This can just be a link to the doc we have about work needed to unblock us, as long as that doc is up to date. Basically can you list the changes you had to make in this PR, and ideally give some indication as to whether there is any follow up work needed? For example as a follow up we should remove all of the dependencies that you've commented out here.

So I'm approving these changes, on the basis that they are the least-bad option we currently have, and acknowledging that some of these changes need follow up work.

File: core/dialog.ts
Comment (cpcallen) [inline core/dialog.ts:123]: Missing `\n` at EOF.

Comment (cpcallen) [inline core/dialog.ts:123]: This seems like avoidable redirection.  Do we need to ensure that we can correctly stub (and unstub) `confirm` even in the face of calls to `setConfirm`?

If not, why not just turn `confirmImplementation` into a property on `TEST_ONLY` and omit `confirmInternal` entirely?

File: core/utils/style.ts
Comment (cpcallen) [inline core/utils/style.ts:302]: Missing `\n` at EOF.  Does your editor have an option analogous to emacs's `require-final-newline`?

**Review by cpcallen — CHANGES_REQUESTED** [review]
I would put the `…Internal` implementations directly after the original function they have hollowed out.  It's more readable, and makes for a cleaner diff.

(Unless you think the original function was in the wrong place, in which case move it to wherever you have put the `…Internal` verison.

File: core/dialog.ts
Comment (BeksOmega) [reply core/dialog.ts:123]: Yeah I thought about just using the `confirmImplementation` here, but I wanted to make it clear that that was an actual feature for dev swapping in their own modal code, whereas we just need something to get tests to run.

Right now I prefer having the extra layer of indirection to indicate that we dealing with a non-ideal situation. I don't want to try to make good final decisions atm.

Comment (cpcallen) [reply core/dialog.ts:123]: That seems reasonable.

File: core/blockly.ts
Comment (BeksOmega) [reply core/blockly.ts]: Turned off clang for these lines.

File: core/clipboard.ts
Comment (BeksOmega) [reply core/clipboard.ts]: Turned it on in my vscode settings. Looked into reporting a bug to clang, but I couldn't figure out which repo was the correct one. Might look into it again some other time.

VSCode settings: https://stackoverflow.com/questions/55374304/add-a-single-empty-line-as-the-last-line-in-the-code-clang-format

File: tests/bootstrap.js
Comment (cpcallen) [inline tests/bootstrap.js:94]: Did you see #6218, and the reason it is still a draft PR and not merged yet?

**Review by cpcallen — COMMENTED** [review]
Whoop whoop!

File: tests/bootstrap.js
Comment (BeksOmega) [reply tests/bootstrap.js:94]: Yeah this was committed accidentally :/ will revert.

Comment (BeksOmega) [reply tests/bootstrap.js:94]: https://github.com/google/blockly/pull/6292

**Pull request merged**
